tiny_cnn/io/caffe/layer_factory_impl.h

```diff
@@ -588,7 +613,49 @@ inline bool layer_match(const std::string& caffetype,
 
 inline std::shared_ptr<layer> create(const caffe::LayerParameter& layer,
                                      const shape_t& in_shape,
-                                     shape_t *out_shape) {
+                                     shape_t* out_shape) {
+    const std::string layer_type = layer.type();
+
+    if (layer_type == "Convolution") {
+        return detail::create_convlayer(layer, in_shape, out_shape);
+    }
+
+    if (layer_type == "InnerProduct") {
+        return detail::create_fullyconnected(layer, in_shape, out_shape);
+    }
+
+    if (layer_type == "Pooling") {
+        return detail::create_pooling(layer, in_shape, out_shape);
+    }
+
+    if (layer_type == "LRN") {
+        return detail::create_lrn(layer, in_shape, out_shape);
+    }
+
+    if (layer_type == "Dropout") {
+        return detail::create_dropout(layer, in_shape, out_shape);
+    }
+
+    if (layer_type == "SoftmaxWithLoss" ||
+        layer_type == "Softmax") {
+        return detail::create_softmax(layer, in_shape, out_shape);
+    }
+
+    if (layer_type == "SigmoidCrossEntropyLoss" ||
+        layer_type == "Sigmoid") {
+        return detail::create_sigmoid(layer, in_shape, out_shape);
+    }
+
+    if (layer_type == "ReLU") {
+        return detail::create_relu(layer, in_shape, out_shape);
+    }
+
+    if (layer_type == "TanH") {
+        return detail::create_tanh(layer, in_shape, out_shape);
+    }
+
+    throw std::runtime_error("layer parser not found");
+
```

Comment:
@nyanp I had to work around this solution since with previous std::function<> doesn't compile in this new branch